caikit_tgis_backend/managed_tgis_subprocess.py

```diff
@@ -0,0 +1,371 @@
+
+        return resp.status_code == 200
+
+    def _create_grpc_client(self):
+        log.debug("Connecting to TGIS at [%s] INSECURE", self._hostname)
+        channel = grpc.insecure_channel(self._hostname)
+        self._tgis_client = generation_pb2_grpc.GenerationServiceStub(channel)
+
+    def _handle_autorecovery(self):
+        """Check if the subprocess might be unhealthy and recover"""
+
+        # if we are booting, do nothing
+        if self._tgis_state == _TGISState.BOOTING:
+            log.debug("Autorecovery skipped because the process is booting")
+            return
+
+        # if the health check passes, do nothing
+        try:
+            if self._health_check():
+                log.debug("Autorecovery skipped because health check is passing")
+                return
+        except requests.exceptions.RequestException as e:
+            log.debug("Autorecovery health check failed: %s", repr(e))
+            # handle recovery below
+
+        # restart the process
+        with self._mutex:
+            # double check that we aren't already booting
+            if self._tgis_state == _TGISState.BOOTING:
+                log.debug(
+                    "Autorecovery (in mutex) skipped because the process is booting"
+                )
+                return
+
+            log.warning("Autorecovery is restarting TGIS")
+            self._launch()
+
+        return
+
+
+class _MockRPCError(grpc.RpcError):
+    """Mocks an RpcError but is generated outside the gRPC code"""
+
+    def __init__(self, code):
+        self._code = code
+
+    def code(self):
+        return self._code
+
+
+# used to create closures over the subprocess that can be called like a
+# client stub, but have a auto-recovering client to the subprocess
+class _WrappedRPC:
+    managed_tgis: ManagedTGISSubprocess
+    rpc_name: str
+
+    def __init__(self, managed_tgis, rpc_name):
+        self.managed_tgis = managed_tgis
+        self.rpc_name = rpc_name
+
+    def __call__(self, request, context=None):
+        # TODO: should wait for the subprocess up until the deadline if booting?
+        if not self.managed_tgis.is_ready():
+            raise _MockRPCError(grpc.StatusCode.UNAVAILABLE)
+
+        try:
+            log.debug3("Calling wrapped RPC [%s]", self.rpc_name)
+            return getattr(self.managed_tgis._tgis_client, self.rpc_name).__call__(
+                request, context
+            )
+        except grpc.RpcError as rpc_error:
+            if rpc_error.code() == grpc.StatusCode.CANCELLED:
+                raise rpc_error
+
+            if rpc_error.code() == grpc.StatusCode.UNAVAILABLE:
+                # trigger a check on the process
+                threading.Thread(
+                    target=self.managed_tgis._handle_autorecovery, daemon=True
+                ).start()
+                raise rpc_error
+
+            exc_chain = RuntimeError("Unhandled RPC error")
```

Comment:
I'm not sure I'm totally following the logic here. Why is `rpc_error` what gets raised instead of `exc_chain`? Also, I think the `error` call will itself `raise`, so the following `raise rpc_error` line will never be hit (unless errors are globally disabled)?

Reply:
I'll take another pass at this section, there's probably some more cases I should be handling and comments to be added.

> Why is rpc_error what gets raised instead of exc_chain?

The idea behind adding using `exc_chain` is to add an extra bit of information to indicate that the wrapped client did not have a branch for the particular kind of error, which also means that no auto-recovery is triggered. For the UNAVAILABLE case, the error is "handled" so I didn't wrap it up in the exception chain. Also, I just noticed that I can include a `root_exception` in the call to `error` instead of doing that manually like I did.

> Also, I think the error call will itself raise, so the following raise rpc_error line will never be hit ?

Ah, yep good point. That reminds me of a question I had that I didn't see from a glance at the `error_handler`:
Can I plug into the error handler's logging without raising? Should I call `_handle_exception_messages` directly?

> (unless errors are globally disabled)

There is a way to cause the `error_handler` to not raise? If so, wouldn't an explicit `return` after it be good to have where flow control is needed?